Mathlib/Init/Data/List/Instances.lean

```diff
@@ -2,15 +2,51 @@
 Copyright (c) 2016 Microsoft Corporation. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Leonardo de Moura
+
+! This file was ported from Lean 3 source module lean_core.init.data.list.instances
+! leanprover-community/lean commit 9af482290ef68e8aaa5ead01aa7b09b7be7019fd
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
 -/
 import Mathlib.Init.Data.List.Lemmas
+
 /-!
 Decidable Instances for `List` not (yet) in `Std`
 -/
 
+universe u
+
+attribute [local simp] List.join List.ret
+
 namespace List
 
-variable {α : Type _} {p : α → Prop} [DecidablePred p]
+instance instMonad : Monad List.{u} where
+  pure := @List.ret
+  bind := @List.bind
+  map := @List.map
+#align list.monad List.instMonad
+
+instance instLawfulMonad : LawfulMonad List.{u} := LawfulMonad.mk'
+  (id_map := map_id)
+  (pure_bind := fun _ _ => List.append_nil _)
+  (bind_assoc := by
+    intro _ _ _ l _ _
+    induction l with
+    | nil => simp [Bind.bind]
+    | cons _ _ ih => simp [Bind.bind] at ih; simp [Bind.bind, ih])
+  (bind_pure_comp := by
+    intro _ _ _ l
+    induction l <;> simp [Bind.bind, Functor.map, pure, List.ret] at *; assumption)
```

Comment:
Can't you prove these with 1-liners like in the old code? Or do you need a heavy import for that?

Reply:
Yes. The old code is depends on lemmas in Data.List.Basic.

Reply:
Yes. The old code is depends on lemmas in Data.List.Basic.

Reply:
Could you please leave a comment explaining how to prove these in 1 line if that import is present?
We should probably split `Data.List.Basic` into smaller pieces. Then this file can have a lightweight import that is useful for proofs like these.

Reply:
I moved some lemmas from `Data.List.Basic`.